mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessageVaultHook.java

```diff
@@ -0,0 +1,129 @@
+ ****************************************************************/
+
+package org.apache.james.vault;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+
+import org.apache.james.core.MailAddress;
+import org.apache.james.core.MaybeSender;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.MetadataWithMailboxId;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.extension.PreDeletionHook;
+import org.apache.james.mailbox.model.MessageMetaData;
+import org.apache.james.mailbox.store.MailboxSessionMapperFactory;
+import org.apache.james.mailbox.store.mail.MessageMapper;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+import org.apache.james.mailbox.store.mail.model.impl.MessageParser;
+import org.apache.james.mime4j.MimeException;
+import org.apache.james.mime4j.codec.DecodeMonitor;
+import org.apache.james.mime4j.dom.Message;
+import org.apache.james.mime4j.dom.address.AddressList;
+import org.apache.james.mime4j.dom.address.Mailbox;
+import org.apache.james.mime4j.dom.address.MailboxList;
+import org.apache.james.mime4j.message.DefaultMessageBuilder;
+import org.apache.james.mime4j.stream.MimeConfig;
+import org.reactivestreams.Publisher;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
+
+import reactor.core.publisher.Flux;
+import reactor.util.function.Tuples;
+
+public class DeletedMessageVaultHook implements PreDeletionHook {
+
+    private final DeletedMessageVault deletedMessageVault;
+    private final MailboxSessionMapperFactory mapperFactory;
+
+    @Inject
+    public DeletedMessageVaultHook(DeletedMessageVault deletedMessageVault, MailboxSessionMapperFactory mapperFactory) {
+        this.deletedMessageVault = deletedMessageVault;
+        this.mapperFactory = mapperFactory;
+    }
+
+    @Override
+    public Publisher<Void> notifyDelete(DeleteOperation deleteOperation) {
+        MailboxSession mailboxSession = deleteOperation.getMailboxSession();
+
+        Stream<DeletedMessage> deletedMessagesStream = deleteOperation.getDeletionMetadataList().stream()
+            .flatMap(Throwing.function(metadataWithMailboxId -> createDeletedMessage(mailboxSession, metadataWithMailboxId)));
+
+        return Flux.fromStream(deletedMessagesStream)
+            .flatMap(deletedMessage -> deletedMessageVault.append(mailboxSession.getUser(), deletedMessage));
+    }
+
+    private Stream<DeletedMessage> createDeletedMessage(MailboxSession mailboxSession, MetadataWithMailboxId metadataWithMailboxId) throws MailboxException {
+        MessageMetaData messageMetaData = metadataWithMailboxId.getMessageMetaData();
+        return mapperFactory.getMessageIdMapper(mailboxSession)
```

Comment:
It is a case where Flux would perform better:

- you could call IO methods into elastic scheduler
- you could call CPU methods (like parseMessge) into parallel scheduler